sdcm/utils/distro.py

```diff
@@ -14,38 +14,98 @@
 import enum
 import logging
 
-
 LOGGER = logging.getLogger(__name__)
 
 
 class DistroError(Exception):
     pass
 
+def static_init(cls):
+    if getattr(cls, "static_init", None):
+        cls.static_init(cls)
+    return cls
 
-# pylint: disable=too-many-public-methods
 @enum.unique
-class Distro(enum.Enum):
-    UNKNOWN = (None, None)
-    CENTOS7 = ("centos", "7")
-    CENTOS8 = ("centos", "8")
-    RHEL7 = ("rhel", "7")
-    RHEL8 = ("rhel", "8")
-    OEL7 = ("ol", "7")
-    OEL8 = ("ol", "8")
-    AMAZON2 = ("amzn", "2")
-    ROCKY8 = ("rocky", "8")
-    DEBIAN8 = ("debian", "8")
-    DEBIAN9 = ("debian", "9")
-    DEBIAN10 = ("debian", "10")
-    DEBIAN11 = ("debian", "11")
-    UBUNTU14 = ("ubuntu", "14.04")
-    UBUNTU16 = ("ubuntu", "16.04")
-    UBUNTU18 = ("ubuntu", "18.04")
-    UBUNTU20 = ("ubuntu", "20.04")
-    UBUNTU21 = ("ubuntu", "21.04")
-    UBUNTU21_10 = ("ubuntu", "21.10")
-    UBUNTU22 = ("ubuntu", "22.04")
-    SLES15 = ("sles", "15")
+class DISTRO_BASE(enum.Enum):
+    RHEL = enum.auto(),
+    DEBIAN = enum.auto(),
+    UNKNOWN = enum.auto(),
+
+#A list of tupples of the form [enum_prefix, os name, versions list (format is either "major.minor" or "major"), boolean - debian_like == True else rehl_like ]
+SUPPORTED_OS = [
+    ("CENTOS", "centos", ["7", "8"], DISTRO_BASE.RHEL),
+    ("RHEL", "rhel",["7", "8"], DISTRO_BASE.RHEL),
+    ("OEL", "ol", ["7", "8"], DISTRO_BASE.RHEL),
+    ("AMAZON", "amzn", ["2"], DISTRO_BASE.RHEL),
+    ("ROCKY", "rocky", ["8"], DISTRO_BASE.RHEL),
+    ("DEBIAN","debian", ["8", "9", "10", "11"], DISTRO_BASE.DEBIAN),
+    ("UBUNTU", "ubuntu", ["14.04", "16.04", "18.04", "20.04", "21.04", "21.10", "22.04"], DISTRO_BASE.DEBIAN),
+    ("SLES","sles", ["15"], DISTRO_BASE.UNKNOWN),
+    ("FEDORA","fedora", ["34","35","36"], DISTRO_BASE.RHEL),
+]
+
+
+enum_data = {
+}
+
+distro_base_mapping = {
+
+}
+
+for enum_prefix, os, versions, distro_base in SUPPORTED_OS:
+
+    if distro_base not in distro_base_mapping:
+        distro_base_mapping[distro_base] = []
+    distro_base_mapping[distro_base].append(os)
+    for version in versions:
+        version_for_member_names = version.replace(".","_")
+        enum_data[enum_prefix + version_for_member_names] = (os, version)
+
+@static_init
```

Comment:
this can be replace by calling `EnumFunctionalMixin.static_init()` once after the code was introduced.

Reply:
I think it is more correct to do it with a decorator. This is more of a concept than a functional thing. In other languages
it is called a static constructor, here is an example from C#:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors

So actually this decorator kind of declares that this class has a static constructor.

However since it is a general concept, I moved the introduction of the decorator into it's own separate commit and moved it to the sdcm.utils.decorators module.